sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala

```diff
@@ -255,59 +255,65 @@ private[sql] trait SQLTestUtilsBase
    * Drops temporary view `viewNames` after calling `f`.
    */
   protected def withTempView(viewNames: String*)(f: => Unit): Unit = {
-    try f finally {
-      // If the test failed part way, we don't want to mask the failure by failing to remove
-      // temp views that never got created.
-      try viewNames.foreach(spark.catalog.dropTempView) catch {
-        case _: NoSuchTableException =>
-      }
-    }
+    tryWithFinally(f)(viewNames.foreach(spark.catalog.dropTempView))
   }
 
   /**
    * Drops global temporary view `viewNames` after calling `f`.
    */
   protected def withGlobalTempView(viewNames: String*)(f: => Unit): Unit = {
-    try f finally {
-      // If the test failed part way, we don't want to mask the failure by failing to remove
-      // global temp views that never got created.
-      try viewNames.foreach(spark.catalog.dropGlobalTempView) catch {
-        case _: NoSuchTableException =>
-      }
-    }
+    tryWithFinally(f)(viewNames.foreach(spark.catalog.dropGlobalTempView))
   }
 
   /**
    * Drops table `tableName` after calling `f`.
    */
   protected def withTable(tableNames: String*)(f: => Unit): Unit = {
-    try f finally {
-      tableNames.foreach { name =>
-        spark.sql(s"DROP TABLE IF EXISTS $name")
-      }
-    }
+    tryWithFinally(f)(tableNames.foreach { name =>
+      spark.sql(s"DROP TABLE IF EXISTS $name")
+    })
   }
 
   /**
    * Drops view `viewName` after calling `f`.
    */
   protected def withView(viewNames: String*)(f: => Unit): Unit = {
-    try f finally {
+    tryWithFinally(f)(
       viewNames.foreach { name =>
         spark.sql(s"DROP VIEW IF EXISTS $name")
       }
-    }
+    )
   }
 
   /**
    * Drops cache `cacheName` after calling `f`.
    */
   protected def withCache(cacheNames: String*)(f: => Unit): Unit = {
-    try f finally {
-      cacheNames.foreach { cacheName =>
-        try uncacheTable(cacheName) catch {
-          case _: AnalysisException =>
+    tryWithFinally(f)(cacheNames.foreach(uncacheTable))
+  }
+
+  /**
+   * Executes the given tryBlock and then the given finallyBlock no matter whether tryBlock throws
+   * an exception. If both tryBlock and finallyBlock throw exceptions, the exception thrown
+   * from the finallyBlock with be added to the exception thrown from tryBlock as a
+   * suppress exception. It helps to avoid masking the exception from tryBlock with exception
+   * from finallyBlock
+   */
+  private def tryWithFinally(tryBlock: => Unit)(finallyBlock: => Unit): Unit = {
```

Comment:
Unfortunately, the proposed one looks like with the existing [tryWithSafeFinally](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/util/Utils.scala#L1347)? Does this has a new feature compared to the existing `tryWithSafeFinally`?

Reply:
You are right. They looks almost exactly the same. And this function was in Spark four years already.. I didn’t do enough research on what we already have in Spark... 

The function I created is redundant. Even the name looks very similar. I will update the test to reuse this Utils.tryWithSafeFinally in SQLTestUtils.